core/include/detray/geometry/index_geometry.hpp

```diff
@@ -125,46 +127,51 @@ class index_geometry {
     using portal_filling_container =
         array_type<vector_type<portal>, e_mask_types>;
 
-    struct object_registry {
-        // Known primitives
-        enum id : unsigned int {
-            e_object_types = 2,
-            e_surface = 0,
-            e_portal = 1,
-            e_any = 1,  // defaults to portal
-            e_unknown = 3,
-        };
-
-        template <typename value_type>
-        static constexpr auto get() {
-            if constexpr (std::is_same_v<value_type, surface>) {
-                return e_surface;
-            }
-            if constexpr (std::is_same_v<value_type, portal>) {
-                return e_portal;
-            }
-            return e_unknown;
-        }
-    };
+    // object type
+    using object_registry_type = object_registry<surface, portal>;
 
     // Volume type
-    using volume_type = volume<object_registry, dindex_range, array_type>;
+    using volume_type = volume<object_registry_type, dindex_range, array_type>;
 
     /** Default constructor */
     index_geometry() = default;
 
+    /** Constructor with vecmem memory resource **/
+    DETRAY_HOST
+    index_geometry(vecmem::memory_resource &resource)
+        : _volumes(&resource), _surfaces(&resource), _portals(&resource) {}
+
+    /** Constructor from index_geometry_data
+     **/
+    template <typename index_geometry_data_t,
+              std::enable_if_t<!std::is_base_of_v<vecmem::memory_resource,
+                                                  index_geometry_data_t>,
+                               bool> = true>
+    DETRAY_DEVICE index_geometry(index_geometry_data_t &geometry_data)
+        : _volumes(geometry_data._volumes_data),
+          _surfaces(geometry_data._surfaces_data),
+          _portals(geometry_data._portals_data) {}
+
     /** @return total number of volumes */
-    const size_t n_volumes() const { return _volumes.size(); }
+    DETRAY_HOST_DEVICE
+    size_t n_volumes() const { return _volumes.size(); }
 
-    /** @return all volumes in the geometry - const access. */
+    /** @return all volumes in the geometry - const access. (const) */
+    DETRAY_HOST_DEVICE
     const auto &volumes() const { return _volumes; }
 
+    /** @return all volumes in the geometry - const access. (non-const) */
+    DETRAY_HOST_DEVICE
+    auto &volumes() { return _volumes; }
```

Comment:
Where is this needed?

Reply:
They are required from `vecmem::get_data(geometry.volumes())`

```c++
    DETRAY_DEVICE index_geometry(index_geometry_data_t &geometry_data)
        : _volumes(geometry_data._volumes_data),
          _surfaces(geometry_data._surfaces_data),
          _portals(geometry_data._portals_data) {}
```

Simply putting const qualifier to `index_geometry_data_t` does not solve it. because const `vecmem::get_data` is defined for const element type:
```c++
/// Helper function creating a @c vecmem::data::vector_view object
template <typename TYPE, typename ALLOC>
VECMEM_HOST data::vector_view<TYPE> get_data(std::vector<TYPE, ALLOC>& vec);

/// Helper function creating a @c vecmem::data::vector_view object
template <typename TYPE, typename ALLOC>
VECMEM_HOST data::vector_view<const TYPE> get_data(
    const std::vector<TYPE, ALLOC>& vec);
```